### DISCUSSION TITLE
BearSSL EC key/cert generation

### DISCUSSION THREAD

**Discussion by soosp** [esp8266/Arduino#8772] [Q&A]
Hi,
Is this possible to generate EC key and (self signed) certificate onboard using BearSSL? There is a br_ec_keygen() function in bearssl_ec.h, but need some example how to to use it.

**Comment by mcspr** [comment]
I see a tool for key generation in the bearssl distribution
https://github.com/earlephilhower/bearssl-esp8266/blob/b024386d461abd1b7b9be3117e2516b7541f1201/tools/skey.c#L433-L460
(also found at *$core_installation_dir*/tools/sdk/ssl/bearssl/tools/skey.c)

Right now, also needs 'Non-32-Bit Access' (`-DNON32XFER_HANDLER`) or 3.x.x SDK set in build settings.
https://gist.github.com/mcspr/d9d457b918ef777a0720e5b12fb70036

It is crashing at this point in the code, so I'd assume `*order` and `order[N]` should be wrapped with `pgm_read_byte`
https://github.com/earlephilhower/bearssl-esp8266/blob/b024386d461abd1b7b9be3117e2516b7541f1201/src/ec/ec_keygen.c#L44
Since the crash occurs after it trying to access `u8 P256_N[];` that is located in flash (at least, for this variant)
https://github.com/earlephilhower/bearssl-esp8266/blob/c0b69dfb837f0027180c72f13f7c90cfafb83c16/src/ec/ec_secp256r1.c#L27
